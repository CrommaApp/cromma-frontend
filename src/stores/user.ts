import { atom } from 'recoil';
import { Keyword } from 'src/types/home/types';

export const userState = atom<{
	isLogin: boolean;
	id: string;
	recentKeywords: Keyword[];
}>({
	key: 'userState',
	default: {
		isLogin: false,
		id: '',
		recentKeywords: [],
	},
});
