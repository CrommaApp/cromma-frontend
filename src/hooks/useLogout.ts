import React from 'react';
import AuthService from '@services/auth/auth-service';
import { useRecoilState, useSetRecoilState } from 'recoil';
import { UserState, userState } from '@stores/user';
import { errorStatusState, successStatusState } from '@stores/status';
import { BASIC_ERROR_MESSAGE, RESPONSE_STATUS_200 } from '@constants/api';

const authService = new AuthService();

type ReturnTypes = [UserState, () => Promise<void>];

const useLogout = (): ReturnTypes => {
	const [user, setUser] = useRecoilState(userState);

	const setErrorStatus = useSetRecoilState(errorStatusState);
	const setSuccessStatus = useSetRecoilState(successStatusState);

	const onLogout = async () => {
		try {
			const result = await authService.logout();

			if (result.statusCode === RESPONSE_STATUS_200) {
				setUser((prev) => {
					return {
						...prev,
						isLogin: false,
						id: '',
					};
				});

				setSuccessStatus({ successMessage: result.message });
			}
		} catch (error) {
			setErrorStatus({
				errorMessage: BASIC_ERROR_MESSAGE,
			});
		}
	};

	return [user, onLogout];
};

export default useLogout;
