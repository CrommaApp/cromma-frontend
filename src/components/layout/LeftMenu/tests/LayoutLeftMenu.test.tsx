import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import LayoutLeftMenu from '@components/layout/LeftMenu';
import { BrowserRouter } from 'react-router-dom';
import { RecoilRoot } from 'recoil';
import * as useLogout from '@hooks/useLogout';
import { UserState } from '@stores/user';

type ReturnType = [UserState, () => Promise<void>];

describe('LayoutLeftMenu Component', () => {
	const showLoginModal = jest.fn();

	const mockedOnLogout = jest.fn();

	const spyOnUseGetAllPosts = (returnData: ReturnType) => {
		const mockedUseLogout = jest.spyOn(useLogout, 'default');
		mockedUseLogout.mockImplementation(() => returnData);

		render(
			<RecoilRoot>
				<BrowserRouter>
					<LayoutLeftMenu showLoginModal={showLoginModal} />
				</BrowserRouter>
			</RecoilRoot>,
		);
	};

	it('Snapshot', () => {
		const component = renderer.create(
			<RecoilRoot>
				<BrowserRouter>
					<LayoutLeftMenu showLoginModal={showLoginModal} />
				</BrowserRouter>
			</RecoilRoot>,
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('Call onLogout() when click logout button', () => {
		spyOnUseGetAllPosts([{ isLogin: true, id: 'test123' }, mockedOnLogout]);

		const logoutButton = screen.getByRole('button', {
			name: '로그아웃',
		});
		userEvent.click(logoutButton);

		expect(mockedOnLogout).toHaveBeenCalled();
	});

	it('Call showLoginModal() when click login button', () => {
		spyOnUseGetAllPosts([{ isLogin: false, id: '' }, mockedOnLogout]);

		const loginButton = screen.getByRole('button', {
			name: '회원가입/로그인',
		});

		userEvent.click(loginButton);

		expect(showLoginModal).toHaveBeenCalled();
	});
});
