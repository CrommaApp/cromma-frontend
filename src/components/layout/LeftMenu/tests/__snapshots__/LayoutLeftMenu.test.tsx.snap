// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayoutLeftMenu Component Snapshot 1`] = `
<div
  className="sc-bdvvtL jvkVjH"
>
  <div>
    <p
      className="sc-gsDKAQ bQCuDM"
    >
      로그인 해주세요
    </p>
    <nav
      className="sc-dkPtRN bDXfHh"
    >
      <ul>
        <li>
          <a
            href="/"
            onClick={[Function]}
          >
            홈
          </a>
        </li>
        <li>
          <a
            href="/post/upload"
            onClick={[Function]}
          >
            게시글 작성
          </a>
        </li>
      </ul>
    </nav>
  </div>
  <button
    onClick={[MockFunction]}
    type="button"
  >
    회원가입/로그인
  </button>
</div>
`;
